File: src/main/java/cspfj/generator/constraint/AbsDiffGenerator.java
package cspfj.generator.constraint;

import java.util.Arrays;
import java.util.SortedSet;
import java.util.TreeSet;

import com.google.common.primitives.Ints;

import cspfj.constraint.semantic.AbsDiff;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.BitVectorDomain;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;

public final class AbsDiffGenerator extends AbstractGenerator {

    public AbsDiffGenerator(final Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        final Variable result;
        final Variable v0;
        final Variable v1;
        if ("absdiff".equals(constraint.description())) {
            result = getSolverVariable(constraint.getVariable(0));
            v0 = getSolverVariable(constraint.getVariable(1));
            v1 = getSolverVariable(constraint.getVariable(2));
        } else {
            throw new IllegalArgumentException("Cannot handle " + constraint);
        }

        int nulls = 0;
        for (Variable v : Arrays.asList(result, v0, v1)) {
            if (v.getDomain() == null) {
                nulls++;
            }
        }

        if (nulls > 1) {
            return false;
        }
        if (nulls == 1) {
            if (result.getDomain() == null) {

                final SortedSet<Integer> values = new TreeSet<Integer>();
                for (int i : v0.getDomain().allValues()) {
                    for (int j : v1.getDomain().allValues()) {
                        values.add(Math.abs(i - j));
                    }
                }
                result.setDomain(new BitVectorDomain(Ints.toArray(values)));

            } else if (v0.getDomain() == null) {

                v0.setDomain(new BitVectorDomain(generateValues(result, v1)));

            } else if (v1.getDomain() == null) {

                v1.setDomain(new BitVectorDomain(generateValues(result, v0)));

            } else {

                throw new IllegalStateException();

            }
        }
        addConstraint(new AbsDiff(result, v0, v1));
        return true;

    }

    private static int[] generateValues(final Variable result,
            final Variable var) {
        final SortedSet<Integer> values = new TreeSet<Integer>();
        for (int i : result.getDomain().allValues()) {
            for (int j : var.getDomain().allValues()) {
                values.add(i + j);
                values.add(i - j);
            }
        }
        return Ints.toArray(values);
    }

}

File: src/main/scala/cspfj/generator/constraint/AbsDiffGenerator.scala
package cspfj.generator.constraint;

import cspfj.constraint.semantic.AbsDiff
import cspfj.problem.{ Variable, Problem, IntDomain }
import cspom.constraint.CSPOMConstraint

final class AbsDiffGenerator(problem: Problem) extends AbstractGenerator(problem) {

  override def generate(constraint: CSPOMConstraint) = {
    require(constraint.description == "absdiff")

    val Seq(result, v0, v1) = constraint.scope.map(cspom2cspfj)

    if (Seq(result, v0, v1).count(_.dom == null) > 1) {
      false
    } else {
      if (result.dom == null) {

        val values = AbstractGenerator.domainFrom(v0, v1, (i, j) => math.abs(i - j))

        result.dom = new IntDomain(values);

      } else if (v0.dom == null) {

        v0.dom = new IntDomain(generateValues(result, v1));

      } else if (v1.dom == null) {

        v1.dom = new IntDomain(generateValues(result, v0));

      }
      addConstraint(new AbsDiff(result, v0, v1));
      true;
    }

  }

  private def generateValues(result: Variable, variable: Variable) = {
    AbstractGenerator.makeDomain(
      AbstractGenerator.cartesian(result, variable, (i, j) => Seq(i + j, i - j)).flatten)
  }
}